### ISSUE TITLE
Push Notification Enhancement: Subscribed Users Request Type

### ISSUE THREAD

**Issue by mjones-oddball** [department-of-veterans-affairs/notification-api#1523]
## User Story - Business Need

The Mobile App Team is looking to create an all user push notification to update them about new features, bug fixes, etc. and drive app engagement. 

There is a push notification option to subscribe users to an SNS topic and call that topic in order to trigger push notifications to all subscribed users. This could be used for as many or as few subscribed users as you want. This will be a new endpoint.

- [x] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

**As a**  Veteran
**I want** to be notified of important app updates
**So that** I can utilize new features and stay up to date on resolutions to issues I care about

### Additional Info and Resources

- Business Intake Form: https://github.com/department-of-veterans-affairs/va.gov-team/issues/67117
- VEText has a test template set up with a couple personalizations for this purpose. Details [here](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Architecture/push-notifications/flagship-mobile-app.md#push-notifications-for-users-subscribed-to-an-sns-topic).
- Sample Request Body for this type of notification:
```
{ "mobile_app": "VA_FLAGSHIP_APP",  
  "topic_sid": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",  
  "template_id": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",  
  "personalisation": {    
    "Message", "New update for the Mobile App..."  
   }  
}
```
- topic_sid = topicSid in VEText
- Including a topic implies it's subscription based, so no recipient_identifier should be provided in this model
- Current [push endpoint](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/v2/notifications/rest_push.py#L21)
- mobile_app has a default in the code, so the field isn't required in the push request. Default value is `VA_FLAGSHIP_APP`

## Engineering Checklist
- [x] Update openapi.yaml with the new route, request body example, and error codes
    - [x] Review with QA for questions/changes before moving on to implementing code
    - [x] Delete openapi-with-push.yaml file
- [x] Create a new push notification route to accept the new request type: `v2/notifications/push/broadcast`
    - Notes: This should use the same auth and validation pattern as the existing push endpoint, the request body just requires a topic sid instead of ICN
- [x] Add validation to ensure all required fields are provided in the request
- [x] Add validation to ensure all fields provided are in the right format
- [x] Update internal postman with the new route
- [x] Update client facing postman with the new route
- [x] Let Strike team know when this ticket is completed so they can schedule updating Portal api docs
## Acceptance Criteria

**NOTE: We should confirm the request is successfully sent to VEText.** If you can see it delivered to your device, great, otherwise we should defer further testing as there's a lot of changes pending - e.g. VEText has to make the preference active, subscribe users to the topic, potentially update the template based on Mobile App Team needs.. etc.

GIVEN you POST a push notification broadcast request to `{{notification-api-url}}/v2/notifications/push/broadcast`

- [x] Successful request
> WHEN you include valid data for mobile_app, topic_sid, template_id, and personalization(s)
> AND VA Notify successfully makes the connection to VEText
> THEN respond with a 201

- [ ] Failed request to VEText
> WHEN you include valid data for mobile_app, topic_sid, template_id, and personalization(s)
> AND VA Notify is not able to make the connection to VEText
> THEN respond with a 503

- [x] Validation error: missing required field
> WHEN you do not include a required field (topic_sid, template_id)
> THEN respond with a 400
> AND include an error message which references the missing required field(s) 

- [x] Validation error: invalid field types
> WHEN you include a field with the wrong format of data
> THEN respond with a 400
> AND include the field and data type expected

## QA Considerations
_For QA to populate. Leave blank if QA is not applicable on this ticket._

VEText has a test template set up with a couple personalizations for this purpose. Details [here](https://github.com/department-of-veterans-affairs/vanotify-team/blob/master/Architecture/push-notifications/flagship-mobile-app.md#push-notifications-for-users-subscribed-to-an-sns-topic).

Reminder that at this stage, the only fields we can validate within Notify are mobile_app and auth (service_id and api key). Template is stored in VEText. We use the Mobile App service for push notifications.

## Out of Scope

-  notification persistence in our database

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1523?planning-poker) with Zenhub @cris-oddball @EvanParish @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by mjones-oddball** [comment]
@k-macmillan will need updates to this ticket based on making it a new endpoint

**Comment by npmartin-oddball** [comment]
Deferring until Monday refinement due to ticket revision

**Comment by mjones-oddball** [comment]
De-prioritizing in favor of resolving the MPI issue, comp and pen, and V3. Business currently doesn't know when this will be worked.

**Comment by kalbfled** [comment]
Thus far I have made modifications to the openapi spec and created a new JSON schema for the new route.

**Comment by kalbfled** [comment]
@cris-oddball I finished my initial work on this ticket, and it is ready for testing.

There's still an open question about which environment to use for testing.

**Comment by cris-oddball** [comment]
When I discussed this earlier with @mjones-oddball she said that you and @k-macmillan should chat about that. The way it is looking it might be best to just point the URL to staging in Dev, Perf and Staging for testing. I can do minimal testing with Postman just to ensure that we get back the expected responses but we will defer a real test with vetext until they are farther down the road with their work.

**Comment by cris-oddball** [comment]
Tentative status after deploying to dev:
- routes appear to be working in a service with a pre-existing push permission
- adding a push permission to a service that does not have one results in a failure (service does not have permission)
- question for @mjones-oddball  re mobile_app as a required field. will ask when she is out of her meeting

**Comment by cris-oddball** [comment]
Log info for push/broadcast sent in dev via Postman with a Happy Path payload. Dev is configured to use the staging vetext endpoint.
```
{ 
    "mobile_app": "VA_FLAGSHIP_APP",  
    "topic_sid": "redacted",  
    "template_id": "redacted",  
    "personalisation": {    
        "message": "Test from VANotify QA"  
   }  
}
```

```
INFO 2024-02-09 16:57:36,942 /app/app/authentication/auth.py:201: API authorised for service <redacted> with api key <redacted>, using client PostmanRuntime/7.36.1

INFO 2024-02-09 16:57:36,944 /app/app/va/vetext/client.py:56: VEText Payload information: {'appSid': '<redacted>', 'topicSid': '<redacted>', 'templateSid': '<redacted>', 'personalization': None}

INFO 2024-02-09 16:57:37,146 /app/app/va/vetext/client.py:63: VEText response: {'success': True, 'statusCode': 200}

10.247.96.243 - - [09/Feb/2024:16:57:37 +0000] "POST /v2/notifications/push/broadcast HTTP/1.1" 201 26 "-" "PostmanRuntime/7.36.1"

```

Note that we did not send the personalization payload to vetext.

**Comment by cris-oddball** [comment]
Update:

- @mjones-oddball is fine with the existing logic for the mobile_app field and updated the AC in this ticket
- after some delay (an hour?) the push permission was recognized as a permission in my QA service and allowed the push to be sent.

The only outstanding question is if we should add the PUSH_NOTIFICATIONS_ENABLED flag to the perf env as a task definition. I asked in our engineering chat.

**Comment by kalbfled** [comment]
The "new problem" (above) was a bug.  I pushed a fix.  Kudos to Cris for catching that.

**Comment by cris-oddball** [comment]
Dave's fix deployed to dev, now the payload includes the personalization value.
```
INFO 2024-02-09 18:13:30,370 /app/app/va/vetext/client.py:56: VEText Payload information: {'appSid': '<redacted>', 'topicSid': '<redacted>', 'templateSid': '<redacted>', 'personalization': {'%MESSAGE%': 'Test from VANotify QA'}}
```

**Comment by cris-oddball** [comment]
PR approved and merged. Sending to Perf for testing.

**Comment by kalbfled** [comment]
@nathanbwright This ticket is still open for QA, but the changes have been merged to master.

**Comment by cris-oddball** [comment]
QA Testing on Perf (manual, not automated)

- FAILED

    WHEN you include valid data for mobile_app, topic_sid, template_id, and personalization(s)
    AND VA Notify successfully makes the connection to VEText
    THEN respond with a 201

```
{ 
    "mobile_app": "VA_FLAGSHIP_APP",  
    "topic_sid": "<redacted>",  
    "template_id": "<redacted>",  
    "personalisation": {    
        "message": "Test from VANotify QA"  
   }  
}
```

Received 503 Response
```
{"message":"Mobile app is not initialized","result":"error"}
```

Beside the logs showing the 503, there are these logs:
```
{
    "asctime": "2024-02-12T16:40:01",
    "application": "app",
    "name": "app",
    "levelname": "WARNING",
    "requestId": null,
    "message": "Missing SID for app: VA_FLAGSHIP_APP",
    "pathname": "/app/app/mobile_app/mobile_app_registry.py",
    "lineno": 25
}

{
    "asctime": "2024-02-12T16:40:01",
    "application": "app",
    "name": "app",
    "levelname": "WARNING",
    "requestId": null,
    "message": "Missing SID for app: VETEXT",
    "pathname": "/app/app/mobile_app/mobile_app_registry.py",
    "lineno": 25
}
```

Dave is looking into it.

**Comment by cris-oddball** [comment]
Additional env vars need to be set for Perf. Awaiting the second PR.

**Comment by cris-oddball** [comment]
Deployed Dave's PR to Perf prior to him merging it to master. Receiving 502 on the Happy Path response

Logs:
```
INFO 2024-02-12 17:45:03,012 /app/app/va/vetext/client.py:56: VEText Payload information: {'appSid': '<redacted>', 'topicSid': '<redacted>', 'templateSid': '<redacted>', 'personalization': {'%MESSAGE%': 'Test from VANotify QA'}}

INFO 2024-02-12 17:45:03,289 /app/app/va/vetext/client.py:63: VEText response: 500

WARNING 2024-02-12 17:45:03,289 /app/app/va/vetext/client.py:76: HTTPError raised sending push notification

WARNING 2024-02-12 17:45:03,290 /app/app/va/vetext/client.py:79: Retryable exception raised with status code: 500

10.247.182.45 - - [12/Feb/2024:17:45:03 +0000] "POST /v2/notifications/push/broadcast HTTP/1.1" 502 72 "-" "PostmanRuntime/7.36.1"

```

**Comment by cris-oddball** [comment]
Re-deploying Dav'e branch to perf with some additions for the basic auth. Will test again when finished.

**Comment by cris-oddball** [comment]
QA Testing in Perf, manual not automated

PUSH BROADCAST

QA Testing on Perf (manual, not automated)

- [ ] Successful request

    WHEN you include valid data for mobile_app, topic_sid, template_id, and personalization(s)
    AND VA Notify successfully makes the connection to VEText
    THEN respond with a 201

```
{ 
    "mobile_app": "VA_FLAGSHIP_APP",  
    "topic_sid": "<redacted>",  
    "template_id": "<redacted>",  
    "personalisation": {    
        "message": "Test from VANotify QA"  
   }  
}
```

201 response from service:
```
{
    "result": "success"
}
```

Logs: 
```
INFO 2024-02-12 21:48:47,814 /app/app/va/vetext/client.py:56: VEText Payload information: {'appSid': '<redacted>', 'topicSid': '<redacted>', 'templateSid': '<redacted>', 'personalization': {'%MESSAGE%': 'Test from VANotify QA'}}

INFO 2024-02-12 21:48:48,024 /app/app/va/vetext/client.py:63: VEText response: {'success': True, 'statusCode': 200}

```

</br>

- UNABLE TO TEST:  Failed request to VEText

    WHEN you include valid data for mobile_app, topic_sid, template_id, and personalization(s)
    AND VA Notify is not able to make the connection to VEText
    THEN respond with a 503

</br>

- [ ] Validation error: missing required field

    WHEN you do not include a required field (topic_sid, template_id)
    THEN respond with a 400
    AND include an error message which references the missing required field(s)

```
{ 
    "mobile_app": "VA_FLAGSHIP_APP",  
    "personalisation": {    
        "message": "Test from VANotify QA"  
   }  
}
```


400 response from service:
```
{
    "errors": [
        {
            "error": "ValidationError",
            "message": "template_id is a required property"
        },
        {
            "error": "ValidationError",
            "message": "topic_sid is a required property"
        }
    ],
    "status_code": 400
}
```
</br>

- [ ] Validation error: invalid field types

    WHEN you include a field with the wrong format of data
    THEN respond with a 400
    AND include the field and data type expected

```
{ 
    "mobile_app": "VA_FLAGSHIP_APP",  
    "topic_sid": 1,  
    "template_id": 2,  
    "personalisation": {    
        "message": "Test from VANotify QA"  
   }  
}
```

400 response from service:
```
{
    "errors": [
        {
            "error": "ValidationError",
            "message": "template_id 2 is not of type string"
        },
        {
            "error": "ValidationError",
            "message": "topic_sid 1 is not of type string"
        }
    ],
    "status_code": 400
}
```
</br>

**Comment by cris-oddball** [comment]
**PASSED QA**

Now just waiting on the new code to be approved and merged.

**Comment by cris-oddball** [comment]
Code approved, merged, already tested so closing out ticket.
